File: cpp-answers/476.数字的补数.cpp
/*
 * @Author: your name
 * @Date: 2020-06-13 20:01:20
 * @LastEditTime: 2020-06-13 20:04:36
 * @LastEditors: your name
 * @Description: In User Settings Edit
 * @FilePath: \undefinede:\leetcode刷题\476.数字的补数.cpp
 */
/*
 * @lc app=leetcode.cn id=476 lang=cpp
 *
 * [476] 数字的补数
 */

// @lc code=start
class Solution
{
public:
    int findComplement(int num)
    {
        int tmp = 1;
        while (tmp < num)
        {
            tmp <<= 1;
            tmp += 1;
        }
        return (tmp ^ num);
    }
};
// @lc code=end

File: java-answers/leetcode/editor/cn/476. 数字的补数.java
class Solution {
    public int findComplement(int num) {
        // 找到最高位的 1 ，然后构造从该位开始到最低位的数字全部都为 1 的数，之后和 num 进行异或
        int highbit = 0;
        for(int i = 0; i <= 30; i++) {
            if((1 << i) > num) {
                highbit = i;
                break;
            }
        }
        int mask = (1 << (highbit)) - 1;
        if(highbit == 0 && num != 1) {
            mask = 0x7fffffff;
        }

        return mask ^ num;
    }
}